File: src/ts/magma/app/compile/define/Definition.ts
import { List } from "../../../../magma/api/collect/list/List";
import { Iterable } from "../../../../magma/api/collect/list/Iterable";
import { Node } from "../../../../magma/app/compile/node/Node";
import { Option } from "../../../../magma/api/option/Option";
import { Some } from "../../../../magma/api/option/Some";
import { Joiner } from "../../../../magma/api/collect/Joiner";
import { TypeCompiler } from "../../../../magma/app/TypeCompiler";
import { RootCompiler } from "../../../../magma/app/RootCompiler";
import { MethodHeader } from "../../../../magma/app/compile/define/MethodHeader";
import { Strings } from "../../../../magma/api/text/Strings";
export class Definition {
	annotations: List<string>;
	modifiers: List<string>;
	typeParams: Iterable<string>;
	type: Node;
	name: string;
	constructor (annotations: List<string>, modifiers: List<string>, typeParams: Iterable<string>, type: Node, name: string) {
		this.annotations = annotations;
		this.modifiers = modifiers;
		this.typeParams = typeParams;
		this.type = type;
		this.name = name;
	}
	findNode(): Node {
		return this.type/*unknown*/;
	}
	toAssignment(): string {
		return "\n\t\tthis." + this.name + " = " + this.name + ";"/*unknown*/;
	}
	generate(): string {
		return this.generateWithAfterName("")/*unknown*/;
	}
	asDefinition(): Option<Definition> {
		return new Some<Definition>(this)/*unknown*/;
	}
	generateWithAfterName(afterName: string): string {
		let joinedNodeParams = this.joinNodeParams()/*unknown*/;
		let joinedModifiers = this.modifiers.iter().map((value: string) => {
			return value + " "/*unknown*/;
		}).collect(new Joiner("")).orElse("")/*unknown*/;
		return joinedModifiers + TypeCompiler.generateBeforeName(this.type) + this.name + joinedNodeParams + afterName + this.generateNode()/*unknown*/;
	}
	generateNode(): string {
		if (this.type.is("var")/*unknown*/){
			return ""/*unknown*/;
		}
		return ": " + TypeCompiler.generateNode(this.type)/*unknown*/;
	}
	joinNodeParams(): string {
		return RootCompiler.joinNodeParams(this.typeParams)/*unknown*/;
	}
	hasAnnotation(annotation: string): boolean {
		return this.annotations.contains(annotation)/*unknown*/;
	}
	removeModifier(modifier: string): MethodHeader {
		return new Definition(this.annotations, this.modifiers.removeNode(modifier), this.typeParams, this.type, this.name)/*unknown*/;
	}
	isNamed(name: string): boolean {
		return Strings.equalsTo(this.name, name)/*unknown*/;
	}
}

File: src/java/magma/app/compile/define/Definition.java
package magma.app.compile.define;

import magma.api.collect.Joiner;
import magma.api.collect.list.Iterable;
import magma.api.collect.list.List;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.api.text.Strings;
import magma.app.RootCompiler;
import magma.app.TypeCompiler;
import magma.app.compile.node.Node;

public record Definition(
        List<String> annotations,
        List<String> modifiers,
        Iterable<String> typeParams,
        Node type,
        String name
) implements MethodHeader, Parameter {
    public Node findNode() {
        return this.type;
    }

    public String toAssignment() {
        return "\n\t\tthis." + this.name + " = " + this.name + ";";
    }

    @Override
    public String generate() {
        return this.generateWithAfterName("");
    }

    public Option<Definition> asDefinition() {
        return new Some<Definition>(this);
    }

    @Override
    public String generateWithAfterName(String afterName) {
        var joinedNodeParams = this.joinNodeParams();
        var joinedModifiers = this.modifiers.iter()
                .map((String value) -> {
                    return value + " ";
                })
                .collect(new Joiner(""))
                .orElse("");

        return joinedModifiers + TypeCompiler.generateBeforeName(this.type) + this.name + joinedNodeParams + afterName + this.generateNode();
    }

    private String generateNode() {
        if (this.type.is("var")) {
            return "";
        }

        return ": " + TypeCompiler.generateType(this.type);
    }

    private String joinNodeParams() {
        return RootCompiler.joinNodeParams(this.typeParams);
    }

    @Override
    public boolean hasAnnotation(String annotation) {
        return this.annotations.contains(annotation);
    }

    @Override
    public MethodHeader removeModifier(String modifier) {
        return new Definition(this.annotations, this.modifiers.removeNode(modifier), this.typeParams, this.type, this.name);
    }

    public boolean isNamed(String name) {
        return Strings.equalsTo(this.name, name);
    }
}
